src/controller.js

```diff
@@ -0,0 +1,126 @@
+        const newPortElement = document.createElement("div");
+        newPortElement.className = "port";
+        newPortElement.dataset.portName = port.name;
+        newPortElement.dataset.portIndex = index;
+        portsElement.appendChild(newPortElement);
+        const portsElementWidth = parseInt(portsElement.style.width, 10);
+        portsElement.style.width = `${portsElementWidth + 256}px`;
+      })
+    }
+
+    renderShip() {
+      const ship = this.ship;
+      const shipPortIndex = ship.itinerary.ports.indexOf(ship.currentPort);
+      const portElement = document.querySelector(`[data-port-index="${shipPortIndex}"]`);
+      if (!portElement) return;
+      const shipElement = document.querySelector("#ship");
+      shipElement.style.top = `${portElement.offsetTop + 32}px`;
+      shipElement.style.left = `${portElement.offsetLeft - 32}px`;
+    }
+
+    setSail() {
+      const ship = this.ship;
+      const currentPortIndex = ship.itinerary.ports.indexOf(ship.currentPort);
+      const nextPortIndex = currentPortIndex + 1;
+      const nextPortElement = document.querySelector(`[data-port-index="${nextPortIndex}"]`);  
+      if (!nextPortElement) {
+        this.renderMessage("End of the line!");
+        return;
+      } else {
+        this.renderMessage(`Now departing ${ship.currentPort.name}`);
+      }
+      const shipElement = document.querySelector("#ship");
+      const sailInterval = setInterval(() => {
+        const shipLeft = parseInt(shipElement.offsetLeft, 10);
+        if (shipLeft === (nextPortElement.offsetLeft - 32)) {
+          document.querySelector("#sailbutton").disabled = false;
+          ship.setSail();
+          ship.dock();
+          clearInterval(sailInterval);
+          this.renderMessage(`Now arriving at ${ship.currentPort.name}`);
+          if (currentPortIndex + 1 === ship.itinerary.ports.length -1) {
+            this.updateDisplay(`Current Port: ${ship.currentPort.name}`, `Next Port: N/A End of the line!`);
+          } else {
+            this.updateDisplay(`Current Port: ${ship.currentPort.name}`, `Next Port: ${ship.itinerary.ports[nextPortIndex + 1].name}`);
+          }
+        }
+        shipElement.style.left = `${shipLeft + 1}px`;
+      }, 20);
+    }
+
+    renderMessage(message) {
+      const messageElement = document.createElement("div");
+      messageElement.id = "message";
+      messageElement.innerHTML = message;
+      const viewport = document.querySelector("#viewport");
+      viewport.appendChild(messageElement); 
+      setTimeout(() => {
+        viewport.removeChild(messageElement)
+      }, 2000);
+    }
+
+    updateDisplay(currentPortInformation, nextPortInformation) {
+      const currentPortUpdate = document.querySelector("#current-port");
+      const nextPortUpdate = document.querySelector("#next-port");
+      currentPortUpdate.innerHTML = currentPortInformation;
+      nextPortUpdate.innerHTML = nextPortInformation;
+    }
```

Comment:
as a follow on from my last comment, I would change this functionality to be 
```
updateDisplay() {
  const currentPort = this.ship.currentPort;
  const ports = this.ship.itinerary.ports;
  const nextPort = ports[ports.indexOf(currentPort) + 1] || { name: "N/A End of the line!" }
    
  const currentPortUpdate = document.querySelector("#current-port");
  const nextPortUpdate = document.querySelector("#next-port");
  currentPortUpdate.innerHTML = `Current Port: ${currentPort.name}`;
  nextPortUpdate.innerHTML = `Next Port: ${nextPort.name}`
}
```

then when you call it you can just call it like `this.updateDisplay();`, this lets you easily call it inside of `addPort` which fixes the issue with the display not properly updating if you add a port while at the previous end of the line